File: java/com/google/copybara/onboard/GeneratorCmdImpl.java
/*
 * Copyright (C) 2024 Google LLC.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.onboard;

import com.google.common.collect.ImmutableList;
import com.google.copybara.CommandEnv;
import com.google.copybara.GeneralOptions;
import com.google.copybara.format.BuildifierOptions;
import com.google.copybara.onboard.core.CannotConvertException;
import com.google.copybara.onboard.core.CannotProvideException;
import com.google.copybara.onboard.core.InputProviderResolver;
import com.google.copybara.onboard.core.template.ConfigGenerator;
import com.google.copybara.util.CommandOutputWithStatus;
import com.google.copybara.util.ExitCode;
import com.google.copybara.util.console.Console;
import com.google.copybara.shell.Command;
import com.google.copybara.shell.CommandException;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Optional;
import javax.annotation.Nullable;

/** Implementation library class for the {@link GeneratorCmd} class. */
public class GeneratorCmdImpl {
  public ExitCode executeCommand(CommandEnv commandEnv, InputProviderResolver resolver)
      throws IOException {
    GeneratorOptions genOpts = commandEnv.getOptions().get(GeneratorOptions.class);
    Console console = commandEnv.getOptions().get(GeneralOptions.class).console();

    ImmutableList<ConfigGenerator> generators = generators();
    Inputs.maybeSetTemplates(generators);
    for (ConfigGenerator generator : generators) {
      // force the generator to initialize its Inputs so tha they are declared in the registry
      var unused = generator.consumes();
    }

    try {
      Optional<Path> path = resolver.resolveOptional(Inputs.GENERATOR_FOLDER);
      if (path.isEmpty()) {
        console.error("Cannot infer a path to place the generated config");
        return ExitCode.COMMAND_LINE_ERROR;
      }
      ConfigGenerator template;
      try {
        template = selectGenerator(resolver, genOpts.template, console);
      } catch (CannotConvertException e) {
        console.error("Cannot infer a template for generating a config. Use --template flag.");
        return ExitCode.COMMAND_LINE_ERROR;
      }
      String config = template.generate(resolver);

      Path configDestination = path.get().resolve("copy.bara.sky");
      if (!Files.exists(configDestination.getParent())) {
        Files.createDirectories(configDestination.getParent());
      }
      Files.writeString(configDestination, config);

      format(commandEnv, configDestination);

      console.infoFmt("%s created", configDestination);

    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      console.error("Interrupted: " + e.getMessage());
      return ExitCode.INTERRUPTED;
    } catch (CannotProvideException e) {
      console.error("Cannot resolve input field: " + e.getMessage());
      return ExitCode.COMMAND_LINE_ERROR;
    }
    return ExitCode.SUCCESS;
  }

  private void format(CommandEnv commandEnv, Path config) throws CannotProvideException {
    GeneralOptions generalOptions = commandEnv.getOptions().get(GeneralOptions.class);
    BuildifierOptions buildifierOptions = commandEnv.getOptions().get(BuildifierOptions.class);
    Command cmd =
        new Command(
            new String[] {
              buildifierOptions.buildifierBin, "-type=bzl", config.toAbsolutePath().toString()
            },
            /* environmentVariables= */ null,
            config.getParent().toFile());
    try {
      CommandOutputWithStatus unused =
          generalOptions.newCommandRunner(cmd).withVerbose(generalOptions.isVerbose()).execute();
    } catch (CommandException e) {
      throw new CannotProvideException("Cannot format generated config " + config, e);
    }
  }

  private ConfigGenerator selectGenerator(
      InputProviderResolver resolver, @Nullable String cliTemplate, Console console)
      throws CannotConvertException, CannotProvideException, InterruptedException {
    ImmutableList<ConfigGenerator> generators = generators();
    if (cliTemplate != null) {
      return Inputs.templateInput().convert(cliTemplate, resolver);
    }
    for (ConfigGenerator generator : generators) {
      if (generator.isGenerator(resolver)) {
        console.info("Using '" + generator.name() + "' template");
        return generator;
      }
    }
    return resolver.resolve(Inputs.templateInput());
  }

  /** A priority ordered lists of templates that can be used */
  protected ImmutableList<ConfigGenerator> generators() {
    return ImmutableList.of(new GitToGitGenerator());
  }
}

File: src/Copybara.Core/Onboard/GeneratorCmdImpl.cs
/*
 * Copyright (C) 2024 Google LLC.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Immutable;

using Copybara.Format;
using Copybara.Onboard.Core;
using Copybara.Onboard.Core.Template;
using Copybara.Util;

using Console = Copybara.Util.Console.Console;

namespace Copybara.Onboard;

/// <summary>
/// Implementation library class for <see cref="GeneratorCmd"/>. Port of
/// <c>com.google.copybara.onboard.GeneratorCmdImpl</c>.
///
/// <para><b>Port note (CLI boundary):</b> takes the resolved <see cref="Options"/> instead of the
/// Copybara.Cli <c>CommandEnv</c>.</para>
/// </summary>
public class GeneratorCmdImpl
{
    public ExitCode ExecuteCommand(Options options, IInputProviderResolver resolver)
    {
        GeneratorOptions genOpts = options.Get<GeneratorOptions>();
        Console console = options.Get<GeneralOptions>().GetConsole();

        IReadOnlyList<IConfigGenerator> generators = Generators();
        Inputs.MaybeSetTemplates(generators);
        foreach (IConfigGenerator generator in generators)
        {
            // Force the generator to initialize its Inputs so that they are declared in the registry.
            _ = generator.Consumes();
        }

        try
        {
            GeneratorFolder? path = resolver.ResolveOptional(Inputs.GeneratorFolder);
            if (path == null)
            {
                console.Error("Cannot infer a path to place the generated config");
                return ExitCode.CommandLineError;
            }

            IConfigGenerator template;
            try
            {
                template = SelectGenerator(resolver, genOpts.Template, console);
            }
            catch (CannotConvertException)
            {
                console.Error("Cannot infer a template for generating a config. Use --template flag.");
                return ExitCode.CommandLineError;
            }

            string config = template.Generate(resolver);

            string configDestination = System.IO.Path.Combine(path.Path, "copy.bara.sky");
            string? parent = System.IO.Path.GetDirectoryName(configDestination);
            if (parent != null && !Directory.Exists(parent))
            {
                Directory.CreateDirectory(parent);
            }

            File.WriteAllText(configDestination, config);

            Format(options, configDestination);

            console.InfoFmt("%s created", configDestination);
        }
        catch (ThreadInterruptedException e)
        {
            console.Error("Interrupted: " + e.Message);
            return ExitCode.Interrupted;
        }
        catch (CannotProvideException e)
        {
            console.Error("Cannot resolve input field: " + e.Message);
            return ExitCode.CommandLineError;
        }

        return ExitCode.Success;
    }

    /// <exception cref="CannotProvideException"/>
    private static void Format(Options options, string config)
    {
        GeneralOptions generalOptions = options.Get<GeneralOptions>();
        BuildifierOptions buildifierOptions = options.Get<BuildifierOptions>();
        string? parent = System.IO.Path.GetDirectoryName(System.IO.Path.GetFullPath(config));
        var cmd =
            new Command(
                new[]
                {
                    buildifierOptions.BuildifierBin, "-type=bzl", System.IO.Path.GetFullPath(config),
                },
                environmentVariables: null,
                parent);
        try
        {
            _ = generalOptions.NewCommandRunner(cmd)
                .WithVerbose(generalOptions.IsVerbose())
                .Execute();
        }
        catch (CommandException e)
        {
            throw new CannotProvideException("Cannot format generated config " + config, e);
        }
    }

    /// <exception cref="CannotConvertException"/>
    /// <exception cref="CannotProvideException"/>
    /// <exception cref="System.Threading.ThreadInterruptedException"/>
    private IConfigGenerator SelectGenerator(
        IInputProviderResolver resolver, string? cliTemplate, Console console)
    {
        IReadOnlyList<IConfigGenerator> generators = Generators();
        if (cliTemplate != null)
        {
            return Inputs.TemplateInput().Convert(cliTemplate, resolver);
        }

        foreach (IConfigGenerator generator in generators)
        {
            if (generator.IsGenerator(resolver))
            {
                console.Info("Using '" + generator.Name + "' template");
                return generator;
            }
        }

        return resolver.Resolve(Inputs.TemplateInput());
    }

    /// <summary>A priority-ordered list of templates that can be used.</summary>
    public virtual IReadOnlyList<IConfigGenerator> Generators() =>
        ImmutableArray.Create<IConfigGenerator>(new GitToGitGenerator());
}
